pkg/controller/noderefutil/providerid.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package noderefutil
+
+import (
+	"errors"
+	"net/url"
+	"path"
+)
+
+var (
+	ErrEmptyProviderID   = errors.New("providerID is empty")
+	ErrInvalidProviderID = errors.New("providerID is not valid")
+)
+
+// ProviderID is a struct representation of a Kubernetes ProviderID.
+// Format: cloudProvider://optional/segments/etc/id
+type ProviderID struct {
+	value *url.URL
+}
+
+// NewProviderID parses the input string and returns a new ProviderID.
+func NewProviderID(id string) (*ProviderID, error) {
+	if id == "" {
+		return nil, ErrEmptyProviderID
+	}
+
+	parsed, err := url.Parse(id)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &ProviderID{
+		value: parsed,
+	}
+
+	if !res.Validate() {
+		return nil, ErrInvalidProviderID
+	}
+
+	return res, nil
+}
+
+// CloudProvider returns the cloud provider portion of the ProviderID.
+func (p *ProviderID) CloudProvider() string {
+	return p.value.Scheme
+}
+
+// ID returns the identifier portion of the ProviderID.
+func (p *ProviderID) ID() string {
+	return path.Base(p.value.Path)
+}
+
+// Equals returns true if both the CloudProvider and ID match.
+func (p *ProviderID) Equals(o *ProviderID) bool {
+	return p.CloudProvider() == o.CloudProvider() && p.ID() == o.ID()
```

Comment:
Wouldn't this result in aws:///<zone>/<awsInstanceId> being != aws:////<awsInstanceId>?

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/aws/instances.go#L48-L52

Reply:
I think we disallow empty IDs for validation, but that should work if you only test Equals

Reply:
oops, markdown killed my comparison...  I meant `aws:////i-xxxxxx` being != `aws:///us-east-1a/i-xxxxxx`

Reply:
ID() will only return the last part of the ProviderID. That comparison you mentioned should be tested here https://github.com/kubernetes-sigs/cluster-api/pull/1011/files#diff-a5350fd6ed4fabde029f67abad7b17a5R110